Dockerfile

```diff
@@ -9,16 +9,36 @@ ARG IOG="false"
 
 RUN DEBIAN_FRONTEND=noninteractive \
  && apt-get update \
- && apt-get -y install curl git jq nix zstd \
+ && apt-get -y install curl gh git grep jq nix zstd \
  && curl -L https://raw.githubusercontent.com/input-output-hk/actions/latest/devx/support/fetch-docker.sh -o fetch-docker.sh \
  && chmod +x fetch-docker.sh \
  && SUFFIX='' \
  && if [ "$MINIMAL" = "true" ]; then SUFFIX="${SUFFIX}-minimal"; fi \
  && if [ "$IOG" = "true" ]; then SUFFIX="${SUFFIX}-iog"; fi \
- && ./fetch-docker.sh input-output-hk/devx $PLATFORM.$COMPILER_NIX_NAME$TARGET_PLATFORM${SUFFIX}-env | zstd -d | nix-store --import | tee store-paths.txt \
- && apt-get -y remove curl git jq nix zstd \
- && apt-get -y autoremove \
- && apt-get -y autoclean
+ && ./fetch-docker.sh input-output-hk/devx $PLATFORM.$COMPILER_NIX_NAME$TARGET_PLATFORM${SUFFIX}-env | zstd -d | nix-store --import | tee store-paths.txt
 
-# `tail -n 2 X | head -n 1` seems a bit fragile way to get `ghc8107-iog-env.sh` derivation path ...
-RUN echo "source $(tail -n 2 store-paths.txt | head -n 1)" >> $HOME/.bashrc
+RUN cat <<EOF >> $HOME/.bashrc
+CACHE_DIR="\$HOME/.cache"
+if [ ! -d "\$CACHE_DIR" ] && [ -n "\$GITHUB_TOKEN" ]; then
+    echo "\$GITHUB_TOKEN" | gh auth login --with-token
+    COMMIT_HASH=\$(git rev-parse HEAD)
+    REPO_URL=\$(git config --get remote.origin.url)
+    if [[ "\$REPO_URL" =~ git@github.com:(.+)/(.+)\.git ]]; then
+        OWNER=\${BASH_REMATCH[1]}
+        REPO=\${BASH_REMATCH[2]}
+    elif [[ "\$REPO_URL" =~ https://github.com/(.+)/(.+).git ]]; then
+        OWNER=\${BASH_REMATCH[1]}
+        REPO=\${BASH_REMATCH[2]}
+    fi
+    if [ -n "\$COMMIT_HASH" ] && [ -n "\$OWNER" ] &&  [ -n "\$REPO" ]; then
+        ARTIFACT_NAME="cache-\$COMMIT_HASH"
+        ARTIFACT_URL=\$(gh api "repos/\$OWNER/\$REPO/actions/artifacts" --jq ".artifacts[] | select(.name==\"\$ARTIFACT_NAME\") | .archive_download_url" | head -n 1)
+        if [ -n "\$ARTIFACT_URL" ]; then
+            curl -L -o "artifact.zstd" -H "Authorization: token \$GITHUB_TOKEN" "\$ARTIFACT_URL"
+            zstd -d "artifact.zstd" --output-dir-flat "\$CACHE_DIR"
+            rm "artifact.zstd"
+        fi
+    fi
+fi
```

Comment:
This needs some discussion/comments on what we try to achieve here and how.
If we created this file with nix instead, we'd get `shellcheck` for free?